src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.cs

```diff
@@ -44,26 +23,29 @@ public static partial class JsonSerializer
             bool finishedSerializing;
             do
             {
-                switch (state.Current.JsonClassInfo.ClassType)
+                WriteStackFrame current = state.Current;
+                switch (current.JsonClassInfo.ClassType)
                 {
                     case ClassType.Enumerable:
-                        finishedSerializing = WriteEnumerable(options, ref writer, ref state);
+                        finishedSerializing = HandleEnumerable(current.JsonClassInfo.ElementClassInfo, options, writer, ref state);
                         break;
                     case ClassType.Value:
-                        finishedSerializing = WriteValue(options, ref writer, ref state.Current);
+                        Debug.Assert(current.JsonPropertyInfo.ClassType == ClassType.Value);
+                        current.JsonPropertyInfo.Write(options, ref current, writer);
+                        finishedSerializing = true;
                         break;
                     case ClassType.Object:
-                        finishedSerializing = WriteObject(options, ref writer, ref state);
+                        finishedSerializing = WriteObject(options, writer, ref state);
                         break;
                     default:
                         Debug.Assert(state.Current.JsonClassInfo.ClassType == ClassType.Unknown);
 
                         // Treat typeof(object) as an empty object.
-                        finishedSerializing = WriteObject(options, ref writer, ref state);
+                        finishedSerializing = WriteObject(options, writer, ref state);
                         break;
                 }
 
-                if (flushThreshold >= 0 && writer.BytesWritten > flushThreshold)
+                if (flushThreshold >= 0 && writer.BytesPending > flushThreshold)
```

Comment:
Note that `writer.BytesWritten` previously showcased the total written (i.e. `BytesCommitted` + whats still pending to be flushed). `BytesPending` now only returns the pending data. Not sure if the semantics were broken before, or now (given that we previously were creating a new `Utf8JsonWriter` every time, but now are not, which likely resolves the issue).

Reply:
BytesPending seems right. Basically we want to call the writer before we fill the buffer too much and force it to grow if the next write pushes it over the edge (can still happen, but shouldn't be common).